### DISCUSSION TITLE
Using notification topics when migrating from Java APNS

### DISCUSSION THREAD

**Discussion by vlads63** [relayrides/pushy#957] [Q&A]
I am migrating to Pushy from Java APNS. In Java APNS to create a notification you only needed a token of the device and payload.

In Pushy you also need a topic. I cannot find any explanation on what this topic is and how to use it. If it was unnecessary in Java APNS why is it mandatory in Pushy? What should I specify as "topic" in my code?

**Comment by jchambers** [comment]
Please see Apple's documentation on ["Sending Notification Requests to APNs"](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/sending_notification_requests_to_apns). It says, in part:

> In general, the topic is your app’s bundle ID/app ID. It can have a suffix based on the type of push notification… To learn more about app ID, see [Register an App ID](https://help.apple.com/developer-account/#/dev1b35d6f83).

Java APNS didn't require a topic generally because it stopped being updated before Apple introduced APNs features that require a topic. In the modern APNs protocol, topics are required more often than not, and so Pushy always requires them to avoid confusing situations where things might just stop working if some seemingly-unrelated part of your setup changes and a topic hasn't been specified.

**Reply by vlads63** [reply]
Thank you for your reply. So App ID is the ID of the application that is sending push notifications, i.e. the application I am converting from Java APNS?

**Reply by jchambers** [reply]
I'm sorry this slipped through the cracks for so long.

> i.e. the application I am converting from Java APNS?

No. It's the [usually] iOS application _receiving_ the push notifications.
